client/trino-jdbc/src/test/java/io/trino/jdbc/TestJdbcExternalAuthentication.java

```diff
@@ -0,0 +1,508 @@
+import com.google.inject.Inject;
+import com.google.inject.Key;
+import io.airlift.configuration.AbstractConfigurationAwareModule;
+import io.airlift.log.Logger;
+import io.airlift.log.Logging;
+import io.trino.client.ClientException;
+import io.trino.client.auth.external.DesktopBrowserRedirectHandler;
+import io.trino.client.auth.external.RedirectException;
+import io.trino.client.auth.external.RedirectHandler;
+import io.trino.plugin.tpch.TpchPlugin;
+import io.trino.server.security.AuthenticationException;
+import io.trino.server.security.Authenticator;
+import io.trino.server.security.ResourceSecurity;
+import io.trino.server.testing.TestingTrinoServer;
+import io.trino.spi.security.Identity;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+import java.io.IOException;
+import java.net.URI;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.time.Duration;
+import java.util.ConcurrentModificationException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.IntSupplier;
+
+import static com.google.common.io.Resources.getResource;
+import static com.google.inject.Scopes.SINGLETON;
+import static io.airlift.jaxrs.JaxrsBinder.jaxrsBinder;
+import static io.trino.jdbc.TestJdbcExternalAuthentication.RedirectHandlerFixture.withHandler;
+import static io.trino.jdbc.TestJdbcExternalAuthentication.TokenPollingErrorFixture.withPollingError;
+import static io.trino.jdbc.TestJdbcExternalAuthentication.WwwAuthenticateHeaderFixture.withWwwAuthenticate;
+import static io.trino.jdbc.TrinoDriverUri.setRedirectHandler;
+import static io.trino.server.security.ResourceSecurity.AccessType.PUBLIC;
+import static io.trino.server.security.ServerSecurityModule.authenticatorModule;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.HttpHeaders.AUTHORIZATION;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON_TYPE;
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN;
+import static javax.ws.rs.core.Response.Status.NOT_FOUND;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@Test(singleThreaded = true)
```

Comment:
maybe add a comment saying this is single threaded because it manuiplates the static redirectio handler